### PR TITLE
pcloud: implement SetModTime and OpenWriterAt

### PR DESCRIPTION
**Pull request by gfelbing**
#### What is the purpose of this change?

This is the second approach on #7896 by implementing OpenWriterAt instead of OpenChunkWriter (#7895), proposed [here](https://github.com/rclone/rclone/issues/7896#issuecomment-2227097288) 

#### Was the change discussed in an issue or in the forum before?

- Issue #7896 

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: fs/operations/copy.go [hunk 1]
```diff
@@ -97,7 +98,12 @@ func (c *copy) checkPartial() (remoteForCopy string, inplace bool, err error) {
 	}
 	// Avoid making the leaf name longer if it's already lengthy to avoid
 	// trouble with file name length limits.
-	suffix := "." + random.String(8) + c.ci.PartialSuffix
+
+	// generate a stable random suffix: use the significant bytes of the hashed remote
+	hashBytes := sha256.Sum256([]byte(c.remote))
```
Comment (ncw) [inline fs/operations/copy.go:106]: Making the suffix depend on the remote name will make it identical for all users. This means that if A and B are uploading at the same time then the resultant file won't be either A or B but a corrupted mix of both.

You could use `fs.Fingerprint(ctx, c.src, true)` to create a string which depends on the source in a reasonably robust fashion. I'm not sure this will solve your problem though.

**Review by ncw — COMMENTED** [review]
I like the SetModTime and OpenWriterAt changes (though I didn't review the code in detail yet).

However I don't like the resume changes. Rclone backends don't resume, and it is very difficult to do and ensure data integrity (how are we 100% sure that the partial file uploaded is the same as the start of the file we are about to upload?).

Can you remove the resuming element from this please and we can get the SetModTime and OpenWriterAt changes in. Then we can discuss resume. There is lots of prior art for discussing resume in the issues!

Thank you :-)

File: fs/operations/copy.go (hunk 1 above)
Comment (gfelbing) [reply fs/operations/copy.go]: Thanks, I'll consider it when we work on the resume part.
I removed all resume-related changes from this PR, including the stable partial filename.

**Comment by gfelbing** [comment]
@ncw , thanks for the first review.
I removed the resume related changes, I see your point of discussing it separately.
This PR would be done with the reduced scope then.
Tests are currently running, I'll upload the result as soon as they are done.

File: backend/pcloud/pcloud.go [hunk 2]
```diff
@@ -326,14 +330,18 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
 	f.features = (&fs.Features{
 		CaseInsensitive:         false,
 		CanHaveEmptyDirectories: true,
+		OpenWriterAt: func(ctx context.Context, remote string, size int64) (fs.WriterAtCloser, error) {
+			return newWriterAt(ctx, remote, f, size)
```
Comment (ncw) [inline backend/pcloud/pcloud.go:334]: Normally we make `OpenWriterAt` a method on `*Fs` - then it can easily read the config etc.

File: backend/pcloud/writer_at.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	ctx    context.Context
+	client *rest.Client
+	fs     *Fs
+	size   int64
+	remote string
+	fd     int64
+	fileID int64
+}
+
+func newWriterAt(
```
Comment (ncw) [inline backend/pcloud/writer_at.go:28]: This should be a method on `*Fs`

**Review by ncw — COMMENTED** [review]
I think that looks great - thank you :-)

Can you make the OpenWriterAt a method on Fs then it will fit with the other implementations and save a few lines of code. You can leave the implementation in `writer_at.go` - that is fine.

File: backend/pcloud/writer_at.go (hunk 3 above)
Comment (gfelbing) [reply backend/pcloud/writer_at.go]: done

File: backend/pcloud/pcloud.go (hunk 2 above)
Comment (gfelbing) [reply backend/pcloud/pcloud.go]: done

**Comment by gfelbing** [comment]
@ncw thanks for the review! I implemented your suggestions.
While trying to run the tests, I ran into one final thing: I need to adjust the default chunk- & buffersize for the pcloud backend.
According to the docs, the backends are supposed to configure their natural chunk sizes themselves.
I tried to look up how others do it (e.g. smb, azurefiles), but I struggled to find the place were they override the values for `--multi-thread-chunk-size` and `--multi-thread-write-buffer-size`. Can you give me a pointer on how this works?

**Comment by ncw** [comment]
> thanks for the review! I implemented your suggestions. 

Great :-)

> While trying to run the tests, I ran into one final thing: I need to adjust the default chunk- & buffersize for the pcloud backend. According to the docs, the backends are supposed to configure their natural chunk sizes themselves. I tried to look up how others do it (e.g. smb, azurefiles), but I struggled to find the place were they override the values for `--multi-thread-chunk-size` and `--multi-thread-write-buffer-size`. Can you give me a pointer on how this works?

The option for configuring natural chunk sizes is in the `OpenChunkWriter` interface. There isn't one in the `OpenWriterAt` interface as it is a more general interface.

Can you explain more about what are good and bad chunk sizes for pcloud and what are the consequences of getting it wrong?

**Comment by gfelbing** [comment]
@ncw I see, then we'd back to OpenChunkWriter :)

Actually, I am more happy with the OpenWriterAt, as it leaves the chunking logic to operations.

It's just the parameters, that need to have better defaults for pcloud:

* Smaller cutoff: the fileops API seems to work way better for medium sized images (10MB and up) already, while the file API works better for smaller sizes, especially when there are lots of them.
* Smaller bufersize: As the fileops API relies on TCP connections for handling file descriptors, it shows that there are more hickups when having longer running requests on the same connection, hence it shouldn't be too large. On the other hand, small many small buffers seem to be very slow. My experiments show that 1-5MB seem to work well.
* Chunksize vs Buffersize: Given a buffersize of 1-5MB and the bottleneck being TCP connections, I don't see an advantage of having a different chunk- than bufersize. I'd set them equally.

That being said, looking at the code: How would you suggest to implement it? IMHO the easiest change would be to make `operations.openChunkWriterFromOpenWriterAt` public and use it for pclouds OpenChunkWriter with backend specific parameters.

**Comment by ncw** [comment]
> Actually, I am more happy with the OpenWriterAt, as it leaves the chunking logic to operations.

:-)

> It's just the parameters, that need to have better defaults for pcloud:

Could you do some experiments and measure throughput for various file sizes? I'd like to see the throughput for the default vs a set of tuned parameters.

We could put some hints in the docs about optimum parameters once we've done this.

> That being said, looking at the code: How would you suggest to implement it? IMHO the easiest change would be to make `operations.openChunkWriterFromOpenWriterAt` public and use it for pclouds OpenChunkWriter with backend specific parameters.

That is a great idea. Do you want to have an experiment with that? We could potentially use that idea to tune the other backends too.

**Comment by gfelbing** [comment]
@ncw , I wrote a script to experiment with the parameters:

- Having test file sets of 32*1M, 16*2M, 8*4M, 4*8M, 2*16M, 1*32M files with random content (each set being 32M large, so a simple time on the rclone will do for throughput)
- Upload each set for every combination of multi thread parameters:
  - cutoff 1M, 2M, 4M, 8M, 16M, 32M
  - buffer size 1M, 2M, 4M, 8M, 16M, 32M
  - chunk size 1M, 2M, 4M, 8M, 16M, 32M
  - streams 1, 2, 4, 8

This makes me upload 162GB of random data, which can take a while...

```bash
#!/usr/bin/env bash

set -ex

if [[ "$1" == "generate" ]]; then
  # generate test files
  for i in {001..32}; do mkdir -p 1M && dd if=/dev/urandom of=1M/$i.rand bs=1M count=1 iflag=fullblock; done
  for i in {001..16}; do mkdir -p 2M && dd if=/dev/urandom of=2M/$i.rand bs=2M count=1 iflag=fullblock; done
  for i in {001..8}; do mkdir -p 4M && dd if=/dev/urandom of=4M/$i.rand bs=4M count=1 iflag=fullblock; done
  for i in {001..4}; do mkdir -p 8M && dd if=/dev/urandom of=8M/$i.rand bs=8M count=1 iflag=fullblock; done
  for i in {001..2}; do mkdir -p 16M && dd if=/dev/urandom of=16M/$i.rand bs=16M count=1 iflag=fullblock; done
  for i in {001..1}; do mkdir -p 32M && dd if=/dev/urandom of=32M/$i.rand bs=32M count=1 iflag=fullblock; done
fi

ALL_SIZES="1M 2M 4M 8M 16M 32M"
ALL_STREAMS="1 2 4 8"

rclone mkdir pcloud:upload-test

for CUTOFF in $ALL_SIZES; do
	for BUFFER_SIZE in $ALL_SIZES; do
		for CHUNK_SIZE in $ALL_SIZES; do
			for STREAMS in $ALL_STREAMS; do
			  for FILES in $ALL_SIZES; do
          rclone delete pcloud:upload-test
          LOG_FILE="$CUTOFF-$BUFFER_SIZE-$CHUNK_SIZE-$STREAMS-$FILES.log"
          { time rclone copy -vv \
            --multi-thread-streams $STREAMS \
            --multi-thread-cutoff $CUTOFF \
            --multi-thread-chunk-size $CHUNK_SIZE \
            --multi-thread-write-buffer-size $BUFFER_SIZE \
            $FILES pcloud:upload-test; } \
            >> $LOG_FILE 2>&1 
			  done
			done
		done
	done
done
```

**Comment by ncw** [comment]
Are the results in? How did it go?

**Comment by gfelbing** [comment]
Not quite, I am on vacation this week, can only run the script when my Starlink is on my RV and stable to produce similar results :)
I'll guess I'll be through next week.

**Comment by gfelbing** [comment]
@ncw script is still ongoing (let it run today for 6h).
Still at cutoff 1M, in the mid off buffer size 2M.
I already created a small script to evaluate the log files, so far it shows a strong correlation (0.69) between file size and time and no correlation between chunk size (-0.01) or concurrent streams (0,07) and time.

Chunk size is no surprise, as all had the same buffer size. That supports my thesis that they should be same.

Concurrent streams is interesting though, I am looking forward wether that changes with more data (especiall different buffer sizes).

I'll give update as soon as the script is finished and post the full data.

**Comment by ncw** [comment]
Great - and good idea measuring the correlation. Interested to hear the full results :-)

**Comment by gfelbing** [comment]
Hey! Just an update: Still working on it.

- I moved the experiment from my local computer to a VPS server, in order to be a/ faster due to higher upload speed and b/ producing more stable results.
- I ran the experiment with the master branch first, as a baseline. This went fine so far.
- With the faster connection, I can reproduce a race condition with pcloud way more often: It seems that there is some sort of server side cache between the fileops API and the regular file API. 

The latter one results in many upload errors - as the checks after the upload uses the file API, while the upload with OpenWriterAt uses the fileops API. Errors are usually size not matching or checksum wrong. 
The upload itself worked properly though - if I look at the deleted files (luckily pcloud has a trash/revisioning system), the uploaded files are correct in size and checksum.
I'll give an update as soon as I narrowed down the exact race condition and try to find a solution.

**Comment by ncw** [comment]
Interesting and fairly common. A sleep and retry is the usual solution here.

**Comment by gfelbing** [comment]
Thanks for the hint, but were to put it? With OpenWriterAt, the check that needs retry happens outside in operations, and within (e.g. during OpenWriterAt.Close) I have no information about the expected checksum of the whole file on which I could retry.

**Comment by ncw** [comment]
> Thanks for the hint, but were to put it? With OpenWriterAt, the check that needs retry happens outside in operations, and within (e.g. during OpenWriterAt.Close) I have no information about the expected checksum of the whole file on which I could retry.

Can you detect the condition in the `Close` method of `OpenWriterAt`?

Maybe by reading the file size / modtime / id / something else and see if it is as you expect? If not retrying?

I'd try just putting a time.Sleep in the `Close` method to see if it can be fixed like that, which at least gives the scale of the problem. We should only use that as a fix as a last resort if we can't find something better though!

**Comment by ncw** [comment]
I'd like to get this in v1.68, is it just the intermittent problem above that is outstanding?

**Comment by gfelbing** [comment]
Yes, that's the only thing which is open.
Currently a bit busy, could check your suggested fix in a few days.

**Comment by ncw** [comment]
No probs.

**Comment by gfelbing** [comment]
> Can you detect the condition in the Close method of OpenWriterAt?

Unfortunately, the writer doesn't get any information about the full file being transmitted (e.g. size, hash). Therefore a check inside is currently not possible.

> Maybe by reading the file size / modtime / id / something else and see if it is as you expect? If not retrying?

Retrying inside the writer isn't really possible as well, as that would undermine the whole purpose of the writer itself, wouldn't it?

> I'd try just putting a time.Sleep in the Close method to see if it can be fixed like that, which at least gives the scale of the problem. We should only use that as a fix as a last resort if we can't find something better though!

It seems that this is sadly the only option currently. At least it does have good effect, the integration tests ([2024-08-17-15-46_test_all.tar.gz](https://github.com/user-attachments/files/16646679/2024-08-17-15-46_test_all.tar.gz)) went through successfully, the experiment above also continues to run.
I'll update the PR with the sleep and an according comment. IMHO this could be merged then, yet I'll still keep on working on a long term solution. But this can easily go into a follow up (e.g. I am currently experimenting with pclouds binary api instead of the http one, to have better control over the underlying connections).

**Comment by ncw** [comment]
> > Can you detect the condition in the Close method of OpenWriterAt?
> 
> Unfortunately, the writer doesn't get any information about the full file being transmitted (e.g. size, hash). Therefore a check inside is currently not possible.
> 
> > Maybe by reading the file size / modtime / id / something else and see if it is as you expect? If not retrying?
> 
> Retrying inside the writer isn't really possible as well, as that would undermine the whole purpose of the writer itself, wouldn't it?

In the Close method, close the file using the special API. Then using the normal API read the size and or checksum and sleep and retry until correct. Only then return from the Close method.

Hopefully that method will work.

You can use the pacer to do the retries and error on too many tries.

**Comment by gfelbing** [comment]
While getting the size/checksum is no problem, how do I determine if it's `correct`? The upload happened via the chunkwriter, which has never seen the source file size/checksum. Am I missing something?

**Comment by gfelbing** [comment]
Ah, forgot that the WriterAt gets the size on creation time. 
That I can use to compare against. 
Updated the PR, looks good on first glance. Tests are currently running.

**Comment by ncw** [comment]
How did the test go?

**Comment by gfelbing** [comment]
Sorry for the cliffhanger :)
The edge-case for unknown (negative sizes) was missing.
Ran the tests again, worked: 
[2024-08-21-195153_test_all.tar.gz](https://github.com/user-attachments/files/16696214/2024-08-21-195153_test_all.tar.gz)

**Comment by ncw** [comment]
Can you fix the lint error? I think you'll need to delete an unused method.

Then I'll merge - thank you :-)

**Comment by gfelbing** [comment]
Updated PR to fix the linter issue (local run succeeded)

**Review by ncw — APPROVED** [review]
Thank you - super 🙂 

Will merge now
